### ISSUE TITLE
Process only one job of a given type at a time across containers

### ISSUE THREAD

**Issue by ajkitson** [Automattic/kue#773]
Is there a way to process only one job of a given type at a time across containers? 

We have a number of job types that we want to process across a handful of containers. It is important that we process the jobs of each type in the order they were added to the queue, so we want only one job from a given job type to be processed at any given time across all the containers. However, we also want to be able to scale horizontally, so while container A is processing one job type, container B can process jobs of another type.

We can use the second parameter on `queue.process` to control concurrency within a container. Is there a way we can control concurrency across containers?

**Comment by behrad** [comment]
Simple answer: No.

You can use a distributed Redis lock in your workers to achieve that, however I can't see why are you running  workers on different containers if your concurrency is ONE! you can simply run any job type on one container :)

**Comment by tobalsgithub** [comment]
We have many job types that we need to process, and we also need to scale horizontally. The best way to do that is to allow all containers to process all job types, and scale up our containers as needed. Restricting certain job types to certain containers requires a tremendous amount of overhead (especially given that our job types are dynamic), and does not necessarily distribute the load equally.

Certainly if we only had 1 job type, and we needed our concurrency to be 1 we would only use 1 container. It's the fact that we have many, many job types that makes this trickier.

Does that help explain the situation better? 

Can you elaborate on using the Redis lock in our processors? Is the idea that in our processing code we'd check to see if the lock was set, and if so return the job to the queue? If not, set the lock and process the job as normal? It's not clear how we could ensure the jobs went back on the queue in the right order.

**Comment by behrad** [comment]
I don't know your deployment considerations, but if you are insisting on deploying an specific job type on several worker processes, I suggest a co-ordinator code who is responsible for calling `.process( typeA )` if and only if a `typeA` redis lock is not acquired by no one else. So you will only one worker started on each container. It also periodically checks lock timeout, so that fresh workers can be started when current master worker of typeA fails and lock times out.

This way you are not handling job re-queues and won't face re-ordering problem.

**Comment by tobalsgithub** [comment]
Yeah, this makes sense. Seems like a redis lock is the way to go here.

Is this something that you would consider a PR for? This redis locking system that you're proposing feels like it could be implemented in the `Worker.start` function: https://github.com/Automattic/kue/blob/master/lib/queue/worker.js#L63

Of course it would have to be an option that was defaulted to off, to maintain the way things work now. But to me, being able to maintain job order across processes seems like something kue should support. On the other hand, if we're the only ones who are trying to do this, maybe we're just weird.

**Comment by behrad** [comment]
A limitation of mutual exclusive processing an item is natural to me, but there are always your environment parameters that define the solution. I still would try to solve this by another deployment structure instead of coding (ensuring each job type is only deployed on one container, not more), and If I couldn't really do that, I would go with distributed co-ordination with locking.

Since this is a really tricky feature, I'd expect you to add it to your own fork, test it, and after a period of working in your deployments, I'd be more than happy to add that to Kue.

Please also note that, currently a redis lock-aware npm module is inside Kue which will be replaced by another one in 1.0 release.

**Comment by tobalsgithub** [comment]
Makes sense to me. Do you know which lock-aware module you'll be replacing the current one with yet?

**Comment by behrad** [comment]
Currently it is `node-redis-warlock` which will be replaced by `https://github.com/mike-marcacci/node-redlock` 
https://github.com/avvs/redislock and https://github.com/errorception/redis-lock are also candidates available on npm.
